operator/standalone/update.go

```diff
@@ -0,0 +1,386 @@
+	u.helmRelease = helmRelease
+	return err
+}
+
+// useReleaseValues unmarshalls and saves release values to be used in subsequent steps
+//
+// This step assumes that the release has been fetched first via fetchHelmRelease.
+func (u *UpdateStandalonePipeline) useReleaseValues(_ context.Context) error {
+	values := helmvalues.V{}
+	err := helmvalues.Unmarshal(u.helmRelease.Spec.ForProvider.Values, &values)
+	u.helmValues = values
+	return err
+}
+
+// applyValuesFromInstance merges the user-defined and -exposed Helm values into the current Helm values map.
+func (u *UpdateStandalonePipeline) applyValuesFromInstance(_ context.Context) error {
+	resources := helmvalues.V{
+		"auth": helmvalues.V{
+			"enablePostgresUser": u.instance.Spec.Parameters.EnableSuperUser,
+			"existingSecret":     getCredentialSecretName(),
+			"database":           u.instance.Name,
+			"username":           u.instance.Name,
+		},
+		"primary": helmvalues.V{
+			"resources": helmvalues.V{
+				"limits": helmvalues.V{
+					"memory": u.instance.Spec.Parameters.Resources.MemoryLimit.String(),
+				},
+			},
+			"persistence": helmvalues.V{
+				"size": u.instance.Spec.Parameters.Resources.StorageCapacity.String(),
+			},
+		},
+		"fullnameOverride": getDeploymentName(),
+		"networkPolicy": helmvalues.V{
+			"enabled": true,
+			"ingressRules": helmvalues.V{
+				"primaryAccessOnlyFrom": helmvalues.V{
+					"enabled": true,
+					"namespaceSelector": helmvalues.V{
+						"kubernetes.io/metadata.name": u.instance.Namespace,
+					},
+				},
+			},
+		},
+	}
+	helmvalues.Merge(resources, &u.helmValues)
+	return nil
+}
+
+// ensureHelmRelease updates the Helm release object
+func (u *UpdateStandalonePipeline) ensureHelmRelease(ctx context.Context) error {
+	chart := &u.config.Spec.HelmReleaseTemplate.Chart
+	helmValues, err := helmvalues.Marshal(u.helmValues)
+	if err != nil {
+		return err
+	}
+	u.helmRelease = &helmv1beta1.Release{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   u.instance.Status.HelmChart.DeploymentNamespace,
+			Labels: getCommonLabels(u.instance.Name),
+		},
+	}
+
+	updateRelease := func() error {
+		u.helmRelease.Spec = helmv1beta1.ReleaseSpec{
+			ForProvider: helmv1beta1.ReleaseParameters{
+				Chart:               helmv1beta1.ChartSpec{Repository: chart.Repository, Name: chart.Name, Version: chart.Version},
+				Namespace:           u.instance.Status.HelmChart.DeploymentNamespace,
+				SkipCreateNamespace: true,
+				SkipCRDs:            true,
+				Wait:                true,
+				ValuesSpec:          helmv1beta1.ValuesSpec{Values: helmValues},
+			},
+			ResourceSpec: crossplanev1.ResourceSpec{
+				ProviderConfigReference: &crossplanev1.Reference{Name: u.config.Spec.HelmProviderConfigReference},
+			},
+		}
+		return nil
+	}
+	_, err = controllerutil.CreateOrUpdate(ctx, u.client, u.helmRelease, updateRelease)
+	return err
+}
+
+func (u *UpdateStandalonePipeline) fetchOperatorConfig(ctx context.Context) error {
```

Comment:
The creation pipeline also has this same function. This should ideally only exist once. Maybe a library with some common functions would make sense?

IMHO one of the pros of the pipeline framework is to easily enable the re-usability and composability so we can get rid of code duplication.

Reply:
Ah sorry didn't see that 🙈